### PR TITLE
Sentry 2.0 [base]

### PR DESCRIPTION
**Pull request by nokitakaze**
Base version for Sentry-PHP 2.0. 
* Namespace Raven instead of prefix "Raven_"
* Removes Raven_Compat
* Renew dependencies in composer.json and .travis.yml

#192, #197 & #424

### CODE REVIEW

**Comment by dcramer** [comment]
fyi https://wordpress.org/about/stats/

I'm ok with us dropping support, but we'd likely want to document "If you're using 5.2, the last version of sentry/sentry that supports it is X"

**Comment by nokitakaze** [comment]
If I'm not mistaken, composer supports downloading these versions of packets what supports needs versions of other packets. So if someone will write 
`composer require getsentry/sentry-php`
It will install old version.

Please **don\'t** merge this pull request at this state (798805b). At first I must add 
`Raven_Client extends \Raven\Client {
}`
otherwise we will crush all packets depends on getsentry/sentry-php.

**Comment by ste93cry** [comment]
Please, there is no point in supporting a version of PHP as old as PHP 5.2 that reached end of life more than [**6 years**](http://php.net/eol.php) ago. Symfony supports PHP `>=5.3.9` in their 2.x branches and `>=5.5.9` for their 3.x branches. Also, I see a lot of PRs about raising the requirements of PHP, but this does not mean that we should move all classes to namespaces because it's a BC. I did not verify if it works, but looking at how the guys of PHPUnit updated their PHP requiremens we should release a new 1.x version that drops PHP 5.2.x support by updating the Composer constraint to **at least** `>=5.6` and remove the custom autoloader `Raven_Autoloader`. Then, all changes related to PSR-4 support, namespaces, etc should be done in a `2.0` branch because they all break BC. On my fork a week ago I pushed the first commit that is similar to the one of this PR, which makes the directory structure compatible with PSR-4 and introduces namespaces. I have a lot of ideas for new 2.x version of this library, but this discussion should be done in a new issue.

**Comment by nokitakaze** [comment]
At my own project I am even more radical https://github.com/nokitakaze/ascetcms/blob/v079/.travis.yml#L3

> all changes related to PSR-4 support, namespaces, etc should be done in a 2.0 branch because they all break BC.

I don't think it\'s necessary. All people calls only Raven_Client. 
At this moment I fix this joking branch. I thought it will be rejected and won't be discussed at all.

**Comment by ste93cry** [comment]
I just noted that even if this library claims to support PHP `5.2`, Travis does not test it. So how can the maintainers be sure their versions work on a so old version?

>I don't think it's necessary. All people calls only Raven_Client. Just give it next minor version

Minor versions can't break BC, this is how SEMVER works. It's not important if users just have to update a line of code, a BC is a BC and cannot be done imho. There is nothing wrong in making a `2.x` release that adds everything we want and let `1.x` as is.

**Comment by nokitakaze** [comment]
- Okay. Now there is file Legacy.php contains old classes.
- And I deleted old "compatible" calls. But I didn't delete Compat.php so this version _should_ works with any code

I think I didn't broke back compatibily. I am **not** sure.

File: composer.json [hunk 1]
```diff
@@ -17,13 +17,13 @@
         "monolog/monolog": "*"
     },
     "require": {
-        "php": ">=5.2.4",
-        "ext-curl": "*"
-    },
-    "suggest": {
+        "php": ">=5.3.2",
```
Comment (ste93cry) [inline composer.json]: I always see `>=5.3.3` as constraint for PHP, so I would do the same here

**Comment by ste93cry** [comment]
When deprecating things I like the approach of Symfony that triggers deprecation notices by putting line of codes like these at the start of a file:

```php
@trigger_error('The '.__NAMESPACE__.'\MyClass class is deprecated since version x.x and will be removed in x.x. Use the NewNameSpace\NewClass class instead.', E_USER_DEPRECATED);
```

Errors can then be catched by frameworks (e.g. Symfony shows deprecations in the profiler) and the users know they are using something that it's going to be removed. However, if you have a single file with all classes I don't know how we could do something like this. I think Symfony guys don't put that lines in the constructors because if someone extended the class then the parent constructor would not be called implicitly, but honestly I'm not sure about this. Please note that I take Symfony as example because it's the framework I use daily, there could be other ways to do the same thing.

File: lib/Raven/Legacy.php [hunk 2]
```diff
@@ -14,7 +16,7 @@ class Raven_Client extends \Raven\Client
 {
 }
 
-/**
+/** @noinspection PhpDeprecationInspection
```
Comment (ste93cry) [inline lib/Raven/Legacy.php]: Given that I was not aware of this tag (thank you for letting me know it 😄), from what I understand it means that IDEs like PHPStorm should not alert about the deprecation of this class. But if the class is deprecated there is no point to not signal it to the users. At least, this is my opinion...

Comment (nokitakaze) [inline lib/Raven/Legacy.php]: No. This tag means new class extends **deprecated** class. The new class (Raven_Compat) is still deprecated and will warn about it. 

I just like green mark.
green
brown-gray

See #409

**Comment by ste93cry** [comment]
>I think I didn't broke back compatibily. I am **not** sure.

I think you broke it. Taking as example people not using Composer, previously they were probably including the `Raven_Autoloader.php` file and then instantiating the class inside it (see [here](https://github.com/getsentry/sentry-php/blob/c4e0a4165a00c3a1186e8554058a29bc197f1d5e/examples/vanilla/index.php#L7) for an example). But with your changes the `Raven_Autoloader` class should be defined inside a new file called `Legacy.php`, that is not included and so it will fail. I said _should_ because I can't find the definition in that file anyway, so it's for sure a BC.

File: lib/Raven/Legacy.php (hunk 2 above)
Comment (ste93cry) [inline lib/Raven/Legacy.php]: I'm not sure to understand correctly what you mean. Also, anyway, why would you add this tag only to this class?

Comment (nokitakaze) [inline lib/Raven/Legacy.php]: My IDEA warns me about extending deprecated class \\Raven\\Compat. I suppress this warning. That's all.

This is the only deprecated class in Raven namespace.

Comment (ste93cry) [inline lib/Raven/Legacy.php]: Thank you for the explanation, now it's much more clear for me. I still don't agree about it, if the inspection warns about a deprecation is for a good reason, in this case both classes are deprecated and should be treated as that. I don't see a good point in having one more line of code just to have a green mark on the IDE, but honestly I don't care too much about it

**Comment by ste93cry** [comment]
You're still breaking compatibility, because if someone did not use the autoloader but manually loaded the files they do not exists anymore. You should keep the old files with the old classes in it, and maybe have a new folder with the new classes. Also, in PSR-4 libraries usually the code is located inside a `src` directory.

**Comment by nokitakaze** [comment]
I can add "require once legacy" into every file. But in this case I must delete E_USER_DEPRECATED

**Comment by ste93cry** [comment]
>But in this case I must delete E_USER_DEPRECATED

Wouldn't moving the `trigger_error` from the `Legacy.php` file to each file fix the problem?

**Comment by nokitakaze** [comment]
We can lose backward compatibility or we can suppress E_USER_DEPRECATED. But not both at the same time. You must pull deprecated classes to files for backward compatibility.

...
Or we can pull spl autoload into any files so it will autoload deprecated classes and trigger E_USER_DEPRECATED.

**Comment by ste93cry** [comment]
>We can lose backward compatibility

This is a no-go, as I said before only major versions can break BC.

>we can suppress E_USER_DEPRECATED. But not both at the same time. You must pull deprecated classes to files for backward compatibility

I do not understand this part, Which problems would arise by moving the `trigger_error` to each old file? If someone uses the new classes he has no problems because no errors are thrown, if he uses one of the old classes then it's correct to get notified of deprecation

**Comment by nokitakaze** [comment]
I don't understand. You want to pull into old files something like this
```php
trigger_error...
require_once 'new/Client.php';
/**
 * @deprecated
 */
class Raven_Client extends \Raven\Client
{
}
```

?

**Comment by ste93cry** [comment]
This is exact, but it would pollute the files with a lot of code just to maintain BC, so I'm not sure all this is worth the work for a `1.x` release

**Comment by nokitakaze** [comment]
The main maintainer should make his views

**Comment by Taluu** [comment]
Just a side note ; upgrading a dependency's version in a minor is fine if it doesn't break the public api... http://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

But +1 on dropping dead versions. Even php 5.3 is not on debian etc anymore, and it's not updated at all (not mentionning that php 5.5 has reached EOL). 

If you want a better graph on what versions are used, you should use the one the maintainer of composer made : https://seld.be/notes/php-versions-stats-2016-2-edition as this is using composer and all (rather than wordpress who is detecting "what is using wordpress", which also counts the "dead" blogs).

**Comment by nokitakaze** [comment]
As I said before my opinion is to drop all old version. Only PHP 5.6+

All old version can use [1.7.0](https://packagist.org/packages/sentry/sentry#dev-master) with #409 & #412. And all new versions of PHP will use 2.0 with dropped EOLed versions.

**Comment by ste93cry** [comment]
>Just a side note ; upgrading a dependency's version in a minor is fine if it doesn't break the public api... 

This is right, but renaming old classes or not being able to load them anymore is breaking the public API. This is why I continue to say that it would be easier to just release a `2.x` version where we can make all adjustment or improvements without any concern about compatibility. A minor version should just be a drop-in replacement that eventually adds new features, but it should never change or remove things

>As I said before my own opinion is to drop all old version. Only PHP 5.6/7.0+

I agree

**Comment by nokitakaze** [comment]
> This branch has conflicts that must be resolved

I will fix conflicts, after @dcramer says he want to merge this pull request. We could use this pull request as a base for #424 (Sentry PHP 2.0).

But I don't know how to test PSR-4 compatibility.

**Comment by ste93cry** [comment]
If this is going to be the base for 2.0, we don't need the `@trigger_error`, nor the autoloadear, nor the `Legacy.php` file

**Comment by nokitakaze** [comment]
Of course

**Comment by ste93cry** [comment]
@nokitakaze Please just keep this PR updated with the latest master changes until it's merged, and if you can squash all commits, but please do not add any commit that does not belongs to PSR-4 switch (e.g. deleting legacy things like the `Compat` class). For that you should open a new PR

**Comment by nokitakaze** [comment]
> please do not add any commit does not belongs to PSR-4 switch

I want @dcramer merging this commit to new branch so anyone else could do another pull requests

**Comment by ste93cry** [comment]
Sure, but removing legacy things like the Rave_Compact class is not in the scope of this PR. The scope of this PR was to move to PSR-4 (you should rename the `lib` directory to `src` and the `test` directory to `tests`) and use Composer, not starting the work for `2.x`

**Comment by nokitakaze** [comment]
>  The scope of this PR was to move to PSR-4

No. This PR was just a **joke**. But after the maintainer approves the idea of version 2.0 (#424) I changed the description and the title of this pull request.

**Comment by ste93cry** [comment]
>I could delete commits with deleting Raven_Compat, but why should I do it?

Because having too many things in a single PR makes review difficult. Removing legacy classes is a thing that should be done separately

**Comment by nokitakaze** [comment]
Ok. I will remove some commits if @dcramer says so.

**Comment by nokitakaze** [comment]
> you should rename the lib directory to src

As I see on http://www.php-fig.org/psr/psr-4/ & https://getcomposer.org/doc/04-schema.md folder "lib" is fine too as long as we save link to it in autoloading section in composer.json

**Comment by ste93cry** [comment]
You're correct, but if you look at most of the other libraries they all tend to use `src` and `tests` as folders

**Comment by ste93cry** [comment]
It looks you missed to add namespaces to the tests classes and rename all them to PSR-4 format. You also renamed only the `test` directory, not to be rude but I found only one of nearly 20 starred repositories I looked at that used `lib`. Also, one more thing: we could add the following to `composer.json`

```json
"autoload-dev": {
    "psr-4": {
        "Raven\\Tests\\": "tests/"
    }
}
```

**Comment by nokitakaze** [comment]
> you missed to add namespaces to the tests

I didn't miss it. I missed it on purpose. I don't think it's necessary to add namespaces to tests

**Comment by ste93cry** [comment]
It's not necessary, but as we're moving to PSR-4 we should move everything to that. Tests classes are source code, and they should be namespaced

**Comment by ste93cry** [comment]
It's not necessary, but as we're moving to PSR-4 we should move everything to that. Tests classes are source code, and they should be namespaced. Just look at [https://getcomposer.org/doc/04-schema.md#autoload-dev](https://getcomposer.org/doc/04-schema.md#autoload-dev) to see that tests are namespaced and they use `src` in all examples

**Comment by nokitakaze** [comment]
Renaming the main folder just because **everyone do it** isn\'t the best thing you can do. Yet again I\'ll do it if @dcramer says so. Same goes for namespaces in test classes

**Comment by ste93cry** [comment]
>Renaming the main folder just because everyone do it isn't the best thing you can do.

If 90% of the people does it, I think we should too because it's a sort of "standard". Also `lib` seems more a directory containing third-parties code, not the _source code_ of the library itself.

**Comment by dcramer** [comment]
Let's start by merging this in. I'll change the branch to 2.0 for now, and we can consider moving to master once we're a bit more prepared with things.

**Comment by dcramer** [comment]
fyi im giving you both commit bit -- the 2.0 and master branches are protected, but you can push to 2.0 as long as the PR is up, and passes status checks (i would aim to go through a brief review on things though)

**Pull request merged**